File: framework/test2/struct/query2d/query2d.cpp
#ifdef __clang__
    #include "soda/clang/stdc++.h"
#else
    #include <bits/stdc++.h>
#endif
#include "soda/leetcode/leetcode.h"
#include "soda/unittest/unittest.h"

using namespace std;
using namespace soda::leetcode;
using namespace soda::unittest;

// step [1]: implement class Solution
// class Solution {};
class NumMatrix {
    int rows, cols;
    vector<vector<int>> matrix;
    vector<vector<int>> bit;
public:
    NumMatrix(vector<vector<int>>& matrix):
        rows{int(matrix.size())+1}, cols{int(matrix[0].size())+1},
        matrix(rows-1, vector<int>(cols-1)),
        bit(rows, vector<int>(cols))
    {
        for (int i = 0; i < rows - 1; ++i) {
            for (int j = 0; j < cols - 1; ++j) {
                update(i, j, matrix[i][j]);
            }
        }
    }
    
    void update(int row, int col, int val) {
        int diff = val - matrix[row][col];
        matrix[row][col] = val;
        for (int i = row+1; i < rows; i += (i & -i)) {
            for (int j = col+1; j < cols; j += (j & -j)) {
                bit[i][j] += diff;
            }
        }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return query(row1, col1) - query(row1, col2+1) - query(row2+1, col1) + query(row2+1, col2+1);
    }

private:
    int query(int r, int c) {
        int res = 0;
        for (int i = r; i > 0; i -= (i & -i)) {
            for (int j = c; j > 0; j -= (j & -j)) {
                res += bit[i][j];
            }
        }
        return res;
    }
};

namespace {
    const auto __ = []() {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);
        return 0;
    }();
}

int main()
{
    // [1] create by class member function
    // Solution su;
    // auto work = WorkFactory::create(su, &Solution::function);
    //
    // [2] or, create by ordinary function
    // auto work = WorkFactory::create(function);

    // auto tester = WorkFactory::createStructTester<NumMatrix,vector<vector<int>>>();
    // tester->withFunction("update", &NumMatrix::update);
    // tester->withFunction("sumRegion", &NumMatrix::sumRegion);
    auto tester = NEW_STRUCT_TESTER(NumMatrix,vector<vector<int>>)
    ADD_FUNCTION(tester, update)
    ADD_FUNCTION(tester, sumRegion)
    auto work = WorkFactory::forStruct(tester);

    // work->setValidator(validate);
    work->setCompareSerial(true);
    work->run();
    delete work;
    return 0;
}

File: test-case/struct/query2d/query2d.go
package main

import (
    . "missingjs.com/soda/leetcode"
    "missingjs.com/soda/unittest"
    "missingjs.com/soda/util"
)

var logger = util.Logger
var _used ListNode

type NumMatrix struct {
    Matrix [][]int
    Bit    [][]int
}


func Constructor(matrix [][]int) NumMatrix {
    rows, cols := len(matrix), len(matrix[0])

    var na NumMatrix
    na.Matrix = make([][]int, rows)
    for i := 0; i < rows; i++ {
        na.Matrix[i] = make([]int, cols)
    }

    na.Bit = make([][]int, rows+1)
    for i := 0; i <= rows; i++ {
        na.Bit[i] = make([]int, cols+1)
    }

    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            na.Update(i, j, matrix[i][j])
        }
    }

    return na
}


func (this *NumMatrix) Update(row int, col int, val int)  {
    diff := val - this.Matrix[row][col]
    this.Matrix[row][col] = val
    for i := row+1; i < len(this.Bit); i += (i & -i) {
        for j := col+1; j < len(this.Bit[0]); j += (j & -j) {
            this.Bit[i][j] += diff
        }
    }
}


func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
    return this.getSum(row2+1, col2+1) + this.getSum(row1, col1) - this.getSum(row2+1, col1) - this.getSum(row1, col2+1)
}


func (this *NumMatrix) getSum(bRow, bCol int) int {
    res := 0
    for i := bRow; i > 0; i -= (i & -i) {
        for j := bCol; j > 0; j -= (j & -j) {
            res += this.Bit[i][j]
        }
    }
    return res
}

func main() {
    // create tester by function
    // work := unittest.CreateWork(FUNCTION)
    // OR create by struct factory
    work := unittest.CreateWorkForStruct(Constructor)
    // work.SetValidator(func(e,r)bool)
    work.CompareSerial = true
    work.Run()
}
